external/java/src/main/java/ch/epfl/dedis/lib/crypto/BlsSig.java

```diff
@@ -0,0 +1,67 @@
+package ch.epfl.dedis.lib.crypto;
+
+import ch.epfl.dedis.lib.crypto.bn256.BN;
+import ch.epfl.dedis.lib.exception.CothorityCryptoException;
+
+import java.math.BigInteger;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.Arrays;
+
+public class BlsSig {
+    private byte[] sig;
+
+    BlsSig(byte[] sig) {
+        this.sig = sig;
+    }
+
+    /**
+     * Verify checks the given BLS signature S on the message m using the public
+     * key X by verifying that the equality e(H(m), X) == e(H(m), x*B2) ==
+     * e(x*H(m), B2) == e(S, B2) holds where e is the pairing operation and B2 is
+     * the base point from curve G2.
+     *
+     * @param msg the signed message .
+     * @param X the public key.
+     * @return true if the verification is successful.
+     */
+    public boolean verify(byte[] msg, Bn256G2Point X) {
+        Bn256G1Point HM = hashToPoint(msg);
+        BN.GT left = HM.pair(X);
+        try {
+            Bn256G1Point s = new Bn256G1Point(sig);
+            if (s.g1 == null) {
+                return false;
+            }
+            BN.GT right = s.pair(new Bn256G2Point(BigInteger.ONE));
+            return Arrays.equals(left.marshal(), right.marshal());
```

Comment:
I was expecting something more like left.equals(right) here. I think I know why it has to be marshalled first (make affine?) but wondering if this means we are doing something wrong design-wise that we can only check equality this way.

Reply:
I don't remember why I did this (this part was implemented some time ago), but I suspect it is because that's the way kyber does it. The `pointGT.Equal` method in kyber basically marshals the two points and then compares them. It's not related to `MakeAffine` which is only used for `G1` and `G2`. I think it should be possible to override the `equals` method in Java and compare the underlying `BigInteger` representation because `GT` doesn't have multiple representations (it is not an elliptic curve point, just a field element).

Reply:
Pushed a commit that added an equals method - 1958fec.